### PR TITLE
 [JENKINS-47043] Introduce exit lifecycle to exit instead of restart

### PR DESCRIPTION
**Pull request by alonbl**
If hudson.lifecycle.ExitLifecycle is selected, jenkins will exit when
restart is requested. Exit code is taken out of
jenkins.model.Jenkins.exitCodeOnRestart, default is 5.

This is usable when there is an external watchdog such as SystemD.

SystemD service example:

   ExecStart=@/usr/bin/java jenkins $JAVA_OPTIONS -DJENKINS_HOME=${JENKINS_HOME} -Dhudson.lifecycle=hudson.lifecycle.ExitLifecycle -Djenkins.model.Jenkins.exitCodeOnRestart=55 -jar $JENKINS_WAR --debug=${JENKINS_DEBUG_LEVEL} --httpPort=${JENKINS_PORT} --httpListenAddress=${JENKINS_LISTEN_ADDRESS} $JENKINS_ARGS

### CODE REVIEW

**Comment by alonbl** [comment]
@daniel-beck what do you think?

File: core/src/main/java/hudson/lifecycle/UnixLifecycle.java
Comment (oleg-nenashev) [inline core/src/main/java/hudson/lifecycle/UnixLifecycle.java]: Same as in #3035

File: core/src/main/java/hudson/lifecycle/UnixLifecycle.java
Comment (oleg-nenashev) [inline core/src/main/java/hudson/lifecycle/UnixLifecycle.java]: unrelated change?

File: core/src/main/java/hudson/lifecycle/UnixLifecycle.java
Comment (oleg-nenashev) [inline core/src/main/java/hudson/lifecycle/UnixLifecycle.java]: The field should be `Integer` then. No need to parse it in runtime and to risk that the logic blows up here in the case of typo in the argument

**Review by oleg-nenashev — CHANGES_REQUESTED** [review]

File: core/src/main/java/hudson/lifecycle/UnixLifecycle.java
Comment (alonbl) [reply core/src/main/java/hudson/lifecycle/UnixLifecycle.java]: ok

File: core/src/main/java/hudson/lifecycle/UnixLifecycle.java
Comment (alonbl) [reply core/src/main/java/hudson/lifecycle/UnixLifecycle.java]: why? we do not need these validation if we use this option, right?

Comment (oleg-nenashev) [reply core/src/main/java/hudson/lifecycle/UnixLifecycle.java]: sorry, NVM. You are right

**Comment by alonbl** [comment]
ping?

**Comment by oleg-nenashev** [comment]
@alonbl GitHub does not send mail notifications if you force to the branch. Will review

**Comment by olivergondza** [comment]
I would be much cleaner to introduce new `Lifecycle` that would exit on restart and let all these watchdog programs to request it explicitly via existing `hudson.lifecycle` property.

**Comment by alonbl** [comment]
ok, will try.

**Comment by alonbl** [comment]
@oleg-nenashev : done

File: core/src/main/java/hudson/lifecycle/ExitLifecycle.java
Comment (olivergondza) [inline core/src/main/java/hudson/lifecycle/ExitLifecycle.java]: What's "Hudson"?

Comment (alonbl) [reply core/src/main/java/hudson/lifecycle/ExitLifecycle.java]: I copied this from here[1], do you prefer I use Jenkins?

[1] https://github.com/jenkinsci/jenkins/blob/master/core/src/main/java/hudson/lifecycle/UnixLifecycle.java#L40

File: core/src/main/java/hudson/lifecycle/ExitLifecycle.java
Comment (oleg-nenashev) [inline core/src/main/java/hudson/lifecycle/ExitLifecycle.java]: Not valid without copyright holder IIRC. IANAL though

File: core/src/main/java/hudson/lifecycle/ExitLifecycle.java
Comment (oleg-nenashev) [inline core/src/main/java/hudson/lifecycle/ExitLifecycle.java]: it would be better to keep the logger field in the top. It's what the Oracle codestyle recommend though not everybody follows it in the core

File: core/src/main/java/hudson/lifecycle/ExitLifecycle.java
Comment (oleg-nenashev) [inline core/src/main/java/hudson/lifecycle/ExitLifecycle.java]: There is no `@Extension` annotation. Should be functionally fine, but in such case the extension point won't be posted in https://jenkins.io/doc/developer/extensions/jenkins-core/#lifecycle

File: core/src/main/java/hudson/lifecycle/ExitLifecycle.java
Comment (oleg-nenashev) [inline core/src/main/java/hudson/lifecycle/ExitLifecycle.java]: Please add `@Restricted(NoExternalUse.class)` so that we will be able to adjust the location/API later if needed

**Review by oleg-nenashev — COMMENTED** [review]
The code generally looks good to me, but I have a concern about the approach (**UPD:** to be explicit, I am fine with the current state if the class is restricted)

* The current lifecycle replaces standard ones if enabled (e. .g. https://github.com/jenkinsci/jenkins/blob/4ab693846ca7a4aa112959a2e92688c3fb9122c3/core/src/main/java/hudson/lifecycle/UnixLifecycle.java)
* It implements only the restart method and suppresses other behaviors
* This installer patch may be useful together with other lifecycles

Maybe it would be better to introduce a new `LifecycleAppender` which would be able to modify existing lifecycles. But maybe it is an overkill.

File: core/src/main/java/hudson/lifecycle/ExitLifecycle.java
Comment (olivergondza) [inline core/src/main/java/hudson/lifecycle/ExitLifecycle.java]: Is there any use-case for specifying the exit code used? I would argue it should be 0, by default at least.

File: core/src/main/java/hudson/lifecycle/ExitLifecycle.java
Comment (olivergondza) [reply core/src/main/java/hudson/lifecycle/ExitLifecycle.java]: "Lifecycle that delegates the responsibility to restart Jenkins to an external entity such as starting process or a service management of some sort."?

File: core/src/main/java/hudson/lifecycle/ExitLifecycle.java
Comment (alonbl) [reply core/src/main/java/hudson/lifecycle/ExitLifecycle.java]: ok, added self.

File: core/src/main/java/hudson/lifecycle/ExitLifecycle.java
Comment (alonbl) [reply core/src/main/java/hudson/lifecycle/ExitLifecycle.java]: Done

File: core/src/main/java/hudson/lifecycle/ExitLifecycle.java
Comment (alonbl) [reply core/src/main/java/hudson/lifecycle/ExitLifecycle.java]: Done, do you want me to submit another patch to fix the other implementations?

File: core/src/main/java/hudson/lifecycle/ExitLifecycle.java
Comment (alonbl) [reply core/src/main/java/hudson/lifecycle/ExitLifecycle.java]: Done, do you want me to submit another patch to fix the other implementations?

File: core/src/main/java/hudson/lifecycle/ExitLifecycle.java
Comment (alonbl) [reply core/src/main/java/hudson/lifecycle/ExitLifecycle.java]: We must provide a way to distinguish between stop and restart. Stop is exit code 0, restart should be something else, I use 5 as default it can be anything, I can put 55 as in the commit message to emphasis it as unique.

Setting it to 0 will result in stop instead of restart, maybe this what you want.

File: core/src/main/java/hudson/lifecycle/ExitLifecycle.java
Comment (alonbl) [reply core/src/main/java/hudson/lifecycle/ExitLifecycle.java]: ok, I copied this from the UnixLifecycle...
Done.

File: core/src/main/java/hudson/lifecycle/ExitLifecycle.java
Comment (oleg-nenashev) [reply core/src/main/java/hudson/lifecycle/ExitLifecycle.java]: would be useful, but it should be in another PR

**Review by oleg-nenashev — APPROVED** [review]
LGTM. @olivergondza are you fine with the PR and the explanation of the default exit code? IMHO whatever works while it is configurable and documented

**Review by olivergondza — APPROVED** [review]

**Comment by daniel-beck** [comment]
@oleg-nenashev 

> It implements only the restart method and suppresses other behaviors

Seems complete to me, UnixLifecycle does the same thing.

**Comment by oleg-nenashev** [comment]
Agreed. I think it is ready to go. @daniel-beck I put it on hold so you have time to review it.

**Comment by daniel-beck** [comment]
🤷‍♂️ Assuming this works and won't need followup options for tons of different things, sure. I'm not familiar enough with the problem domain to say one way or the other. Perhaps @rtyler or @olblak can chime in.

Would it be useful to have a linked PR in the Docker image or native installers to apply this option?

**Comment by oleg-nenashev** [comment]
> Would it be useful to have a linked PR in the Docker image or native installers to apply this option?

It would require a more careful review by maintainers. IIUC https://github.com/jenkinsci/docker does not behave in a compatible way with proper failover. It would require external monitors/failover logic in Kubernetes, Docker Swarm/Compose or whatever

**Comment by daniel-beck** [comment]
@oleg-nenashev So how do we know this feature as implemented actually makes sense?

**Comment by alonbl** [comment]
Hi,

Per systemd, it is fully supported with:
```
Restart=on-success
SuccessExitStatus=5
RestartPreventExitStatus=0
```

Docker wrapper script can be:
```
while true; do
    java....
    status=$?
    [ $status -eq 5 ] || exit $status
done
```

k8s is bit behind, but can enjoy the same as docker.
https://github.com/kubernetes/kubernetes/issues/31147

Regards,
Alon

**Pull request merged**
